### PR TITLE
Nested elements

### PR DESCRIPTION
**Pull request by titusfortner**
Ok, so this is where things start getting interesting...

`#cache=` is supposed to allow for distinguishing between
1. > This element is defined as a wrapping of a Selenium Element and can not be relocated
2. > This element can be relocated by the provided Selector, but the Selenium Element is already located, so we add it as cached.

This PR builds the selector when Element is initialized rather than when it is located.
This allows us to determine on initialization if a single XPath can be created for nested elements, and allows us to bypass recursion during location.

### CODE REVIEW

**Review by p0deje — COMMENTED** [review]
I don't fully follow the changes, so let me withdraw from reviewing. I fully trust your judgement :)

File: lib/watir/elements/element.rb
Comment (jkotests) [inline lib/watir/elements/element.rb]: Sentence ends abruptly.

File: lib/watir/elements/radio.rb
Comment (jkotests) [inline lib/watir/elements/radio.rb]: What is different about `Radio` that requires this? I'm still trying to work through the code, so maybe I'm just missing something. Just seems weird that other places that you added `#driver_scope` were in the `Locator`, where as this is the element itself.

Comment (titusfortner) [reply lib/watir/elements/radio.rb]: Getting `RadioSet` to work was a pain. All that code made sense when I created it, but it was very confusing digging back into it. :-D I'm guessing I needed to have it here to get that to work, but I don't recall. I'll try removing it and see what happens. I should add more comments for these things in general.

**Comment by jkotests** [comment]
@titusfortner , is there still value in reviewing this or should we wait until it can be merged with the overhaul (#825)?

**Comment by titusfortner** [comment]
Hold off, I got halfway through a rebase before getting pulled away.

**Comment by titusfortner** [comment]
@jkotests ok now everything is passing with the latest code merged in

**Comment by titusfortner** [comment]
Ugh. `#merge` and `#merge!` are different. This one is temporarily no longer ready for review. :(

**Comment by titusfortner** [comment]
@jkotests It works!!!

This is based on #826; we can do them at the same time or separately.

This won't work for IFrames, Tables, or Adjacent location, which are probably the 3 main use cases, but for the ever-popular Page Sections, this should be a huge performance bump.

Also, I kind of don't want to touch location again in 2019. :)

**Review by jkotests — APPROVED** [review]

**Pull request merged**
